circleci: failed tests: TestLeaseConcurrent

The following test appears to have failed:

[#23143](https://circleci.com/gh/cockroachdb/cockroach/23143):

```
I160927 20:36:29.047059 1539 storage/replica.go:3295  [s1,r1:/M{in-ax}] gossiping cluster id {"395d5243-e026-46cd-b307-dcdbcbb8adaf"} from store 1, range 1
I160927 20:36:29.048959 1531 storage/replica.go:3317  [s1,r1:/M{in-ax}] gossiping sentinel from store 1, range 1
I160927 20:36:29.049332 1531 storage/replica.go:3324  [s1,r1:/M{in-ax}] gossiping first range from store 1, range 1: [{1 1 1}]
I160927 20:36:29.049827 1539 storage/replica.go:3317  [s1,r1:/M{in-ax}] gossiping sentinel from store 1, range 1
I160927 20:36:29.050150 1539 storage/replica.go:3324  [s1,r1:/M{in-ax}] gossiping first range from store 1, range 1: [{1 1 1}]
panic: (*basictracer.errAssertionFailed) (0x1afc1a0,0xc4201dbd70) [recovered]
	panic: (*basictracer.errAssertionFailed) (0x1afc1a0,0xc4201dbd70)

goroutine 1531 [running]:
panic(0x1afc1a0, 0xc4201dbd70)
	/usr/local/go/src/runtime/panic.go:500 +0x1ae
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Recover(0xc42007f5f0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:177 +0xd8
panic(0x1afc1a0, 0xc4201dbd70)
	/usr/local/go/src/runtime/panic.go:458 +0x271
github.com/opentracing/basictracer-go.(*spanImpl).maybeAssertSanityLocked(0xc42028f560)
	/go/src/github.com/opentracing/basictracer-go/debug.go:27 +0x44a
github.com/opentracing/basictracer-go.(*spanImpl).Lock(0xc42028f560)
	/go/src/github.com/opentracing/basictracer-go/debug.go:19 +0x4d
github.com/opentracing/basictracer-go.(*spanImpl).Log(0xc42028f560, 0x0, 0x0, 0x0, 0xc4203b4c30, 0x48, 0x0, 0x0)
	/go/src/github.com/opentracing/basictracer-go/span.go:106 +0xab
github.com/opentracing/basictracer-go.(*spanImpl).LogEvent(0xc42028f560, 0xc4203b4c30, 0x48)
	/go/src/github.com/opentracing/basictracer-go/span.go:94 +0xaf
github.com/cockroachdb/cockroach/util/log.eventInternal(0x7f84c5ca93a0, 0xc420329920, 0x5e0100, 0x1cf4274, 0x1b, 0xc420269058, 0x2, 0x2)
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc42007f5f0, 0xc4201a8e40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:189 +0x74

goroutine 1 [chan receive]:
testing.(*T).Run(0xc420292300, 0x1ce9caa, 0x13, 0x1fd4b90, 0x9cef01)
	/usr/local/go/src/testing/testing.go:647 +0x56e
testing.RunTests.func1(0xc420292300)
	/usr/local/go/src/testing/testing.go:793 +0xba
testing.tRunner(0xc420292300, 0xc420045b38)
	/usr/local/go/src/testing/testing.go:610 +0xca
testing.RunTests(0x1fd6868, 0x28f1c40, 0x14b, 0x14b, 0xc420111278)
	/usr/local/go/src/testing/testing.go:799 +0x4bb
testing.(*M).Run(0xc420045ef0, 0xc420253ce0)
	/usr/local/go/src/testing/testing.go:743 +0x130
github.com/cockroachdb/cockroach/storage_test.TestMain(0xc420045ef0)
	/go/src/github.com/cockroachdb/cockroach/storage/main_test.go:57 +0x287
main.main()
	github.com/cockroachdb/cockroach/storage/_test/_testmain.go:724 +0x1b6

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1

goroutine 7 [chan receive]:
github.com/cockroachdb/cockroach/util/log.(*loggingT).flushDaemon(0x2c3b400)
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:1032 +0x85
created by github.com/cockroachdb/cockroach/util/log.init.1
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:613 +0xc3

goroutine 1484 [runnable]:
sync.runtime_Semacquire(0xc42007f61c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*WaitGroup).Wait(0xc42007f610)
	/usr/local/go/src/sync/waitgroup.go:131 +0xbf
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Stop(0xc42007f5f0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:365 +0x20a
github.com/cockroachdb/cockroach/storage.(*testContext).Stop(0xc4201d8930)
	/go/src/github.com/cockroachdb/cockroach/storage/replica_test.go:260 +0x4f
github.com/cockroachdb/cockroach/storage.TestLeaseConcurrent.func1(0xc4204f2180, 0x1)
	/go/src/github.com/cockroachdb/cockroach/storage/replica_test.go:1682 +0xa18
--
testing.tRunner(0xc4204f2180, 0x1fd4b90)
	/usr/local/go/src/testing/testing.go:610 +0xca
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x530

goroutine 1529 [chan receive]:
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc42051fce0)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:390 +0x64
created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:391 +0xb61
FAIL	github.com/cockroachdb/cockroach/storage	10.757s
=== RUN   TestBatchBasics
I160927 20:36:41.098732 8 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160927 20:36:41.101289 8 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.01s)
=== RUN   TestBatchRepr
I160927 20:36:41.117073 23 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160927 20:36:41.119211 23 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
--- PASS: TestBatchRepr (0.02s)
=== RUN   TestBatchGet
I160927 20:36:41.128152 10 storage/engine/rocksdb.go:356  opening in memory rocksdb instance

```
Please assign, take a look and update the issue accordingly.

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -1627,8 +1627,7 @@ func makeIDKey() storagebase.CmdIDKey {
 //   which case the other returned values are zero.
 func (r *Replica) proposeRaftCommand(
 	ctx context.Context, ba roachpb.BatchRequest,
-) (
-	chan roachpb.ResponseWithError, func() bool, error) {
+) (chan roachpb.ResponseWithError, func() bool, error) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	if r.mu.destroyed != nil {

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -1620,14 +1620,18 @@ func TestLeaseConcurrent(t *testing.T) {
 				}
 				go func() {
 					wg.Wait()
+					tc.rng.mu.Lock()
+					defer tc.rng.mu.Unlock()
 					if withError {
+						// When we complete the command, we have to remove it from the map;
+						// otherwise its context (and tracing span) may be used after the
+						// client cleaned up.
+						delete(tc.rng.mu.pendingCmds, cmd.idKey)
 						cmd.done <- roachpb.ResponseWithError{
 							Err: roachpb.NewErrorf(origMsg),
 						}
 						return
 					}
-					tc.rng.mu.Lock()
-					defer tc.rng.mu.Unlock()
 					if err := defaultProposeRaftCommandLocked(tc.rng, cmd); err != nil {
 						panic(err) // unlikely, so punt on proper handling
 					}

